### DISCUSSION TITLE
Anyone got experience with A) Desktop Emulation and B) Microphones?

### DISCUSSION THREAD

**Discussion by billy8119** [pschatzmann/arduino-audiokit#38] [Q&A] [answered]
Ultimately, I want to use the esp32-a1s audio kit to do all this, but it's a long road, so I'm trying to use a desktop emulation example to develop and prototype this.  So I took Phil Schatzmann's excellent desktop "generator" example, and tried to initialize "in", which definitely doesn't seem to work:
#include "AudioTools.h"
#include "AudioLibs/PortAudioStream.h"

uint16_t sample_rate=44100;
uint8_t channels = 1;                                     // The stream will have 2 channels 

PortAudioStream out;                                        // On desktop we use 
PortAudioStream in;
PaError ret;

StreamCopy copier(out, in); // copy in to out

// Arduino Setup
void setup(void) {
  Serial.begin(115200);
  AudioLogger::instance().begin(Serial, AudioLogger::Info);

  // open output
  auto config = out.defaultConfig();
  auto incfg = in.defaultConfig();
  config.sample_rate = sample_rate;
  config.channels = channels;
  config.bits_per_sample = sizeof(int16_t)*8;
  
  incfg.sample_rate = sample_rate;
  incfg.channels = channels;
  incfg.bits_per_sample = sizeof(int16_t)*8;

  
  ret = out.begin(config);
  
  // Setup MIC?
  printf("Trying to set up microphone\n");
  ret = in.begin(incfg);
}
// Arduino loop  
void loop() {
  if (out)
    copier.copy();
}
Which compiles fine, but, alas:
ALSA lib pcm_direct.c:2127:(_snd_pcm_direct_new) unable to create IPC semaphore
ALSA lib setup.c:547:(add_elem) Cannot obtain info for CTL elem (MIXER,'AC97 2ch->4ch Copy Switch',0,0,0): No such file or directory
ALSA lib pcm.c:2664:(snd_pcm_open_noupdate) Unknown PCM cards.pcm.center_lfe
ALSA lib pcm.c:2664:(snd_pcm_open_noupdate) Unknown PCM cards.pcm.side
ALSA lib pcm.c:2664:(snd_pcm_open_noupdate) Unknown PCM cards.pcm.surround21
ALSA lib pcm.c:2664:(snd_pcm_open_noupdate) Unknown PCM cards.pcm.surround21
ALSA lib setup.c:547:(add_elem) Cannot obtain info for CTL elem (MIXER,'AC97 2ch->4ch Copy Switch',0,0,0): No such file or directory
ALSA lib pcm.c:2664:(snd_pcm_open_noupdate) Unknown PCM cards.pcm.surround41
ALSA lib pcm.c:2664:(snd_pcm_open_noupdate) Unknown PCM cards.pcm.surround50
ALSA lib pcm.c:2664:(snd_pcm_open_noupdate) Unknown PCM cards.pcm.surround51
ALSA lib pcm.c:2664:(snd_pcm_open_noupdate) Unknown PCM cards.pcm.surround71
ALSA lib setup.c:547:(add_elem) Cannot obtain info for CTL elem (PCM,'IEC958 Playback PCM Stream',0,0,0): No such file or directory
ALSA lib setup.c:547:(add_elem) Cannot obtain info for CTL elem (PCM,'IEC958 Playback PCM Stream',0,0,0): No such file or directory
ALSA lib setup.c:547:(add_elem) Cannot obtain info for CTL elem (PCM,'IEC958 Playback PCM Stream',0,0,0): No such file or directory
ALSA lib pcm.c:2664:(snd_pcm_open_noupdate) Unknown PCM cards.pcm.hdmi
ALSA lib pcm.c:2664:(snd_pcm_open_noupdate) Unknown PCM cards.pcm.hdmi
ALSA lib pcm.c:2664:(snd_pcm_open_noupdate) Unknown PCM cards.pcm.modem
ALSA lib pcm.c:2664:(snd_pcm_open_noupdate) Unknown PCM cards.pcm.modem
ALSA lib pcm.c:2664:(snd_pcm_open_noupdate) Unknown PCM cards.pcm.phoneline
ALSA lib pcm.c:2664:(snd_pcm_open_noupdate) Unknown PCM cards.pcm.phoneline
ALSA lib pcm_oss.c:397:(_snd_pcm_oss_open) Cannot open device /dev/dsp
ALSA lib pcm_oss.c:397:(_snd_pcm_oss_open) Cannot open device /dev/dsp
ALSA lib confmisc.c:160:(snd_config_get_card) Invalid field card
ALSA lib pcm_usb_stream.c:482:(_snd_pcm_usb_stream_open) Invalid card 'card'
ALSA lib confmisc.c:160:(snd_config_get_card) Invalid field card
ALSA lib pcm_usb_stream.c:482:(_snd_pcm_usb_stream_open) Invalid card 'card'
ALSA lib pcm_direct.c:2127:(_snd_pcm_direct_new) unable to create IPC semaphore
Trying to set up microphone
[E] PortAudioStream.h : 162 - PortAudio error: Stream is stopped

ALSA lib pcm.c:8568:(snd_pcm_recover) underrun occurred
[I] AudioCopy.h : 121 - StreamCopy::copy 1024 -> 1024 -> 1024 bytes - in 1 hops
[E] PortAudioStream.h : 162 - PortAudio error: Stream is stopped

ALSA lib pcm.c:8568:(snd_pcm_recover) underrun occurred
[E] PortAudioStream.h : 144 - PortAudio error: Output underflowed
ALSA lib pcm.c:8568:(snd_pcm_recover) underrun occurred
[E] PortAudioStream.h : 144 - PortAudio error: Output underflowed
[I] AudioCopy.h : 259 - try write - 2 
ALSA lib pcm.c:8568:(snd_pcm_recover) underrun occurred
[E] PortAudioStream.h : 144 - PortAudio error: Output underflowed

This repeats, of course.
Looks like my feeble attempt to open sound-as-input ... just.. didn't work... 
Anyone have any clues?
Thanks in Advance!

**Comment by pschatzmann** [comment, marked as answer]
I have never tried: but you would use just one PortAudioStream object for input and output. In the config you would set intput = true. Then you should be able to read and write. This is similar to I2S .

I also just added the defaultConfig() method that accepts a mode parameter - like in AudioKitStream and I2Stream. So you can  now call defaultConfig(RXTX_MODE);

**Comment by billy8119** [comment]
Like this? 
#include "AudioTools.h"
#include "AudioLibs/PortAudioStream.h"

uint16_t sample_rate=44100;
uint8_t channels = 1;                                     // The stream will have 2 channels 
PortAudioStream aud;                                        // On desktop we use 
StreamCopy copier(aud,aud); // copy in to out

// Arduino Setup
void setup(void) {
  Serial.begin(115200);
  AudioLogger::instance().begin(Serial, AudioLogger::Info);

  // open output
  auto config = aud.defaultConfig();  
  config.sample_rate = sample_rate;
  config.channels = channels;
  config.bits_per_sample = sizeof(int16_t)*8;
  config.is_input=true;
  aud.begin(config);
}

// Arduino loop  
void loop() {
  if (aud)
    copier.copy();
}

Works, (thank-you very much!) but there are two issues: 
A) underruns all the time 
ALSA lib setup.c:547:(add_elem) Cannot obtain info for CTL elem (MIXER,'AC97 2ch->4ch Copy Switch',0,0,0): No such file or directory
ALSA lib pcm.c:2664:(snd_pcm_open_noupdate) Unknown PCM cards.pcm.center_lfe
ALSA lib pcm.c:2664:(snd_pcm_open_noupdate) Unknown PCM cards.pcm.side
ALSA lib pcm.c:2664:(snd_pcm_open_noupdate) Unknown PCM cards.pcm.surround21
ALSA lib pcm.c:2664:(snd_pcm_open_noupdate) Unknown PCM cards.pcm.surround21
ALSA lib setup.c:547:(add_elem) Cannot obtain info for CTL elem (MIXER,'AC97 2ch->4ch Copy Switch',0,0,0): No such file or directory
ALSA lib pcm.c:2664:(snd_pcm_open_noupdate) Unknown PCM cards.pcm.surround41
ALSA lib pcm.c:2664:(snd_pcm_open_noupdate) Unknown PCM cards.pcm.surround50
ALSA lib pcm.c:2664:(snd_pcm_open_noupdate) Unknown PCM cards.pcm.surround51
ALSA lib pcm.c:2664:(snd_pcm_open_noupdate) Unknown PCM cards.pcm.surround71
ALSA lib setup.c:547:(add_elem) Cannot obtain info for CTL elem (PCM,'IEC958 Playback PCM Stream',0,0,0): No such file or directory
ALSA lib setup.c:547:(add_elem) Cannot obtain info for CTL elem (PCM,'IEC958 Playback PCM Stream',0,0,0): No such file or directory
ALSA lib setup.c:547:(add_elem) Cannot obtain info for CTL elem (PCM,'IEC958 Playback PCM Stream',0,0,0): No such file or directory
ALSA lib pcm.c:2664:(snd_pcm_open_noupdate) Unknown PCM cards.pcm.hdmi
ALSA lib pcm.c:2664:(snd_pcm_open_noupdate) Unknown PCM cards.pcm.hdmi
ALSA lib pcm.c:2664:(snd_pcm_open_noupdate) Unknown PCM cards.pcm.modem
ALSA lib pcm.c:2664:(snd_pcm_open_noupdate) Unknown PCM cards.pcm.modem
ALSA lib pcm.c:2664:(snd_pcm_open_noupdate) Unknown PCM cards.pcm.phoneline
ALSA lib pcm.c:2664:(snd_pcm_open_noupdate) Unknown PCM cards.pcm.phoneline
Cannot connect to server socket err = No such file or directory
Cannot connect to server request channel
jack server is not running or cannot be started
JackShmReadWritePtr::~JackShmReadWritePtr - Init not done for -1, skipping unlock
JackShmReadWritePtr::~JackShmReadWritePtr - Init not done for -1, skipping unlock
Cannot connect to server socket err = No such file or directory
Cannot connect to server request channel
jack server is not running or cannot be started
JackShmReadWritePtr::~JackShmReadWritePtr - Init not done for -1, skipping unlock
JackShmReadWritePtr::~JackShmReadWritePtr - Init not done for -1, skipping unlock
ALSA lib pcm_oss.c:397:(_snd_pcm_oss_open) Cannot open device /dev/dsp
ALSA lib pcm_oss.c:397:(_snd_pcm_oss_open) Cannot open device /dev/dsp
ALSA lib pulse.c:242:(pulse_connect) PulseAudio: Unable to connect: Connection refused

ALSA lib pulse.c:242:(pulse_connect) PulseAudio: Unable to connect: Connection refused

ALSA lib confmisc.c:160:(snd_config_get_card) Invalid field card
ALSA lib pcm_usb_stream.c:482:(_snd_pcm_usb_stream_open) Invalid card 'card'
ALSA lib confmisc.c:160:(snd_config_get_card) Invalid field card
ALSA lib pcm_usb_stream.c:482:(_snd_pcm_usb_stream_open) Invalid card 'card'
Cannot connect to server socket err = No such file or directory
Cannot connect to server request channel
jack server is not running or cannot be started
JackShmReadWritePtr::~JackShmReadWritePtr - Init not done for -1, skipping unlock
JackShmReadWritePtr::~JackShmReadWritePtr - Init not done for -1, skipping unlock
[E] PortAudioStream.h : 162 - PortAudio error: Stream is stopped

[I] AudioCopy.h : 121 - StreamCopy::copy 1024 -> 1024 -> 1024 bytes - in 1 hops
[I] AudioCopy.h : 121 - StreamCopy::copy 1024 -> 1024 -> 1024 bytes - in 1 hops
[I] AudioCopy.h : 121 - StreamCopy::copy 1024 -> 1024 -> 1024 bytes - in 1 hops
[E] PortAudioStream.h : 144 - PortAudio error: Output underflowed
[I] AudioCopy.h : 259 - try write - 2 
[I] AudioCopy.h : 121 - StreamCopy::copy 1024 -> 1024 -> 1024 bytes - in 2 hops
[I] AudioCopy.h : 121 - StreamCopy::copy 1024 -> 1024 -> 1024 bytes - in 1 hops
[E] PortAudioStream.h : 144 - PortAudio error: Output underflowed
[I] AudioCopy.h : 259 - try write - 2 
[I] AudioCopy.h : 121 - StreamCopy::copy 1024 -> 1024 -> 1024 bytes - in 2 hops
[I] AudioCopy.h : 121 - StreamCopy::copy 1024 -> 1024 -> 1024 bytes - in 1 hops
[E] PortAudioStream.h : 144 - PortAudio error: Output underflowed
[I] AudioCopy.h : 259 - try write - 2 
[I] AudioCopy.h : 121 - StreamCopy::copy 1024 -> 1024 -> 1024 bytes - in 2 hops
[I] AudioCopy.h : 121 - StreamCopy::copy 1024 -> 1024 -> 1024 bytes - in 1 hops
[E] PortAudioStream.h : 144 - PortAudio error: Output underflowed
[I] AudioCopy.h : 259 - try write - 2 
[I] AudioCopy.h : 121 - StreamCopy::copy 1024 -> 1024 -> 1024 bytes - in 2 hops
[I] AudioCopy.h : 121 - StreamCopy::copy 1024 -> 1024 -> 1024 bytes - in 1 hops
[E] PortAudioStream.h : 144 - PortAudio error: Output underflowed
[I] AudioCopy.h : 259 - try write - 2 

and 
B) 500 mSec latency between mic input and audio output.  Tips? Thoughts? 
C) Also, for some reason, I can only run this as root. Even though I've chmod 777 /dev/snd/*  eshrug.  I can debug as root, it's just weird.

**Comment by pschatzmann** [comment]
I haven't used this much so I can't really help. 
For the underlows maybe changing the log level to Warning might help.

Just to give a heads up: this will work perfectly in the AudioKit: just use aud.defaultConfig(RXTX_MODE);

**Comment by billy8119** [comment]
Well, it's for both platformio and your libraries, I mean, I'm hoping to just use desktop stuff for the short term, this ultimately ends up on the esp32.  Desktop = faster compile .. thanks for that! 
Your hint of using the same instance for Portaudiostream unlocked opening the sound card, thanks.  I set the sample rate to 22050 for now, it'll do.  Latency is still stratospheric though, but maybe it will be less on the esp32.  Now I'm off to play with your fabulous filters! Yay!
TY!

**Reply by billy8119** [reply]
Also thanks for the pointer about the config.is_input=true equivalence.  That'll be handy.

**Comment by pschatzmann** [comment]
Just a word of caution: The floating point operations on a microcontroller are much slower than on a desktop, so you will be limited in the size of the filters. Therefore you need to keep them as small as possible!

**Reply by billy8119** [reply]
Got advice? Is there some way to use 16-bit ints or something?
Ultimately it's going to be distortion + tremolo, but with sinewave rather than triangle wave.  Trying to make Dalek sounds

**Reply by pschatzmann** [reply]
Have a look at the available [AudioEffects](https://pschatzmann.github.io/arduino-audio-tools/html/classaudio__tools_1_1_audio_effect.html)
The available TTS engines also give quite some robotic sound...

If you really need fast Filters on the ESP32 there is a DSP library written by Espressif in assempler which will outperform my generic implementation.
